cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -159,8 +160,17 @@ public AppMetadataStore(Table table, CConfiguration cConf, AtomicBoolean upgrade
   }
 
   @Override
-  protected <T> T deserialize(byte[] serialized, Type typeOfT) {
-    return GSON.fromJson(Bytes.toString(serialized), typeOfT);
+  protected <T> T deserialize(MDSKey key, byte[] serialized, Type typeOfT) {
+    if (RunRecordMeta.class.equals(typeOfT)) {
+      RunRecordMeta meta = GSON.fromJson(Bytes.toString(serialized), RunRecordMeta.class);
```

Comment:
can you explain this? Does the Json not have the correct program/run ID?

Reply:
The `ProgramRunId` is transient, hence is not serialized when stored into MDS (for backward compatibility and also not to duplicate what's in the row key already). Maybe we can reduce the temporary object creation by using a custom type adapter if necessary.

Reply:
I think it's fine. If I had read this from bottom to top, I would have seen the changes down there first :)